exercises/react/src/example.c

```diff
@@ -0,0 +1,210 @@
+{
+   struct child *child = r->first;
+   while (child) {
+      struct cb *cb = child->cell->cb;
+      while (cb) {
+         struct cb *next_cb = cb->next;
+         free(cb);
+         cb = next_cb;
+      }
+      free(child->cell);
+
+      struct child *next = child->next;
+      free(child);
+      child = next;
+   }
+   free(r);
+}
+
+#define SUCCESS 1
+#define FAIL 0
+
+static int add_child(struct reactor *r, struct cell *cell)
+{
+   struct child *child = calloc(1, sizeof(struct child));
+   if (!child) {
+      return FAIL;
+   }
+   child->cell = cell;
+   if (!r->first) {
+      r->first = child;
+   } else {
+      r->last->next = child;
+   }
+   r->last = child;
+   return SUCCESS;
+}
+
+struct cell *create_input_cell(struct reactor *r, int initial_value)
+{
+   struct cell *c = calloc(1, sizeof(struct cell));
+   if (!c) {
+      return NULL;
+   }
+   if (add_child(r, c) != SUCCESS) {
+      free(c);
+      return NULL;
+   }
+   c->reactor = r;
+   c->kind = kind_input;
+   c->value = initial_value;
+   return c;
+}
+
+struct cell *create_compute1_cell(struct reactor *r, struct cell *input,
+                                  compute1 compute)
+{
+   struct cell *c = calloc(1, sizeof(struct cell));
+   if (!c) {
+      return NULL;
+   }
+   if (add_child(r, c) != SUCCESS) {
+      free(c);
+      return NULL;
+   }
+   c->reactor = r;
+   c->kind = kind_compute1;
+   c->input1 = input;
+   c->compute1 = compute;
+   c->value = compute(cell_value(input));
+   return c;
+}
+
+struct cell *create_compute2_cell(struct reactor *r, struct cell *input1,
+                                  struct cell *input2, compute2 compute)
+{
+   struct cell *c = calloc(1, sizeof(struct cell));
+   if (!c) {
+      return NULL;
+   }
+   if (add_child(r, c) != SUCCESS) {
+      free(c);
+      return NULL;
+   }
+   c->reactor = r;
+   c->kind = kind_compute2;
+   c->input1 = input1;
+   c->input2 = input2;
+   c->compute2 = compute;
+   c->value = compute(cell_value(input1), cell_value(input2));
+   return c;
+}
+
+int cell_value(struct cell *c)
+{
+   return c->value;
+}
+
+static void propagate(struct cell *c)
+{
+   int new_value;
+   switch (c->kind) {
+   case kind_compute1:
+      new_value = c->compute1(cell_value(c->input1));
+      break;
+   case kind_compute2:
+      new_value = c->compute2(cell_value(c->input1), cell_value(c->input2));
+      break;
+   default:
+      return;
+   }
+
+   if (new_value != c->value) {
+      c->value = new_value;
+      for (struct cb * cb = c->cb; cb; cb = cb->next) {
+         cb->f(cb->obj, new_value);
+      }
+   }
+
+}
+
+void cell_value_set(struct cell *c, int new_value)
+{
+   c->value = new_value;
+   struct reactor *r = c->reactor;
+
+   // We take the very naive route of updating all cells.
+   // Traversing the tree and only updating the cells dependent on the just-changed cell is possible,
+   // but requires much more memory management to make each cell aware of its dependents.
```

Comment:
I initially tried to take the path mentioned (only update the necessary cells) and got pretty mired down in memory management. I still do intend to attempt such a solution in my submission to the website, but I may need a lot more time to do it. If I complete it and it looks reasonable, I can then propose it be the example solution.

In general, the tests do not check for performance... they **could**, if we had a compute function that keeps track of how often it is called and then enforce that it is only called a certain number of times (and not called if its input cell doesn't change). If I added such a test, I would probably deal with it in this solution by adding a dirty bit on each cell that tracks whether to call its compute function again. Of course, it is still iterating through the entire list of cells, but it's not possible to test that aspect of performance.

In general it seems hard to test performance and/or implementation details like that.

There's a larger question of "What standard must example solutions be held to? Surely they should be good style and idiomatic for the language. But Must they be efficient?" I'm not sure any track has come up with a definitive answer, but I see:

* https://github.com/exercism/xhaskell#example-solution: It doesn't need to be perfect or very elegant. But it should be efficient enough for the test suite to finish in only a few seconds.
* https://github.com/exercism/xgo#example-solutions: there is less need for this code to be a model of style, expression and readability, or to use the best algorithm. Examples can be plain, simple, concise, even naïve, as long as they are correct.

I do predict that the more efficient solution will be have more complex code, so I biased toward keeping it simple for the example solution since it may be the case that track maintainers may have to maintain/make changes in the future.